include/unifex/std_concepts.hpp

```diff
@@ -0,0 +1,324 @@
+
+  template <typename T>
+  UNIFEX_CONCEPT_FRAGMENT(
+    _totally_ordered,
+      requires(detail::_cref_t<T> t, detail::_cref_t<T> u) //
+      (
+        requires(detail::_boolean_testable<decltype(t < u)>),
+        requires(detail::_boolean_testable<decltype(t > u)>),
+        requires(detail::_boolean_testable<decltype(u <= t)>),
+        requires(detail::_boolean_testable<decltype(u >= t)>)
+      ));
+  template <typename T>
+  UNIFEX_CONCEPT
+    totally_ordered =
+      equality_comparable<T> &&
+      UNIFEX_FRAGMENT(unifex::_totally_ordered, T);
+
+  template <typename T, typename U>
+  UNIFEX_CONCEPT_FRAGMENT(
+    _totally_ordered_with,
+      requires(detail::_cref_t<T> t, detail::_cref_t<U> u) //
+      (
+        requires(detail::_boolean_testable<decltype(t < u)>),
+        requires(detail::_boolean_testable<decltype(t > u)>),
+        requires(detail::_boolean_testable<decltype(t <= u)>),
+        requires(detail::_boolean_testable<decltype(t >= u)>),
+        requires(detail::_boolean_testable<decltype(u < t)>),
+        requires(detail::_boolean_testable<decltype(u > t)>),
+        requires(detail::_boolean_testable<decltype(u <= t)>),
+        requires(detail::_boolean_testable<decltype(u >= t)>)
+      ));
+      //&& totally_ordered<
+      //   common_reference_t<detail::_cref_t<T>, detail::_cref_t<U>>>
+  template <typename T, typename U>
+  UNIFEX_CONCEPT
+    totally_ordered_with =
+      totally_ordered<T> &&
+      totally_ordered<U> &&
+      equality_comparable_with<T, U> &&
+      //common_reference_with<detail::_cref_t<T>, detail::_cref_t<U>> &&
+      UNIFEX_FRAGMENT(unifex::_totally_ordered_with, T, U);
+
+  ////////////////////////////////////////////////////////////////////////////////////////////
+  // Object concepts
+  ////////////////////////////////////////////////////////////////////////////////////////////
+
+  template <typename T>
+  UNIFEX_CONCEPT
+    destructible =
+      std::is_nothrow_destructible_v<T>;
+
+  template <typename T, typename... Args>
+  UNIFEX_CONCEPT
+    constructible_from =
+      destructible<T> &&
+      UNIFEX_IS_CONSTRUCTIBLE(T, Args...);
+
+  template <typename T>
+  UNIFEX_CONCEPT
+    default_constructible =
+      constructible_from<T>;
+
+  template <typename T>
+  UNIFEX_CONCEPT
+    move_constructible =
+      constructible_from<T, T> &&
+      convertible_to<T, T>;
+
+  template <typename T>
+  UNIFEX_CONCEPT_FRAGMENT(
+    _copy_constructible,
+      requires()(0) &&
+      constructible_from<T, T &> &&
+      constructible_from<T, T const &> &&
+      constructible_from<T, T const> &&
+      convertible_to<T &, T> &&
+      convertible_to<T const &, T> &&
+      convertible_to<T const, T>);
+  template <typename T>
+  UNIFEX_CONCEPT
+    copy_constructible =
+      move_constructible<T> &&
+      UNIFEX_FRAGMENT(unifex::_copy_constructible, T);
+
+  template <typename T>
+  UNIFEX_CONCEPT_FRAGMENT(
+    _move_assignable,
+      requires()(0) &&
```

Comment:
What does this syntax mean?

Reply:
It's the same as `requires() { 0; }`. It is an empty requirement. The way `UNIFEX_CONCEPT_FRAGMENT` is currently defined, a `requires` clause is required and must be first. I could make a `requires` clause optional at the expense of a huge pile of PP gunk. It wouldn't be hard, it's just unpleasant.